Framework/LiveData/inc/MantidLiveData/Kafka/IKafkaStreamDecoder.h

```diff
@@ -0,0 +1,206 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2016 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+#ifndef MANTID_LIVEDATA_IKAFKASTREAMDECODER_H_
+#define MANTID_LIVEDATA_IKAFKASTREAMDECODER_H_
+
+#include "MantidAPI/SpectraDetectorTypes.h"
+#include "MantidDataObjects/EventWorkspace.h"
+#include "MantidLiveData/Kafka/IKafkaBroker.h"
+#include "MantidLiveData/Kafka/IKafkaStreamSubscriber.h"
+
+#include <atomic>
+#include <condition_variable>
+#include <mutex>
+#include <thread>
+#include <unordered_map>
+
+namespace Mantid {
+namespace LiveData {
+
+/**
+Kafka stream decoder interface. Handles (implements) all thread synchronization
+functionality for accessing the data stream and processing data.
+*/
+class DLLExport IKafkaStreamDecoder {
+public:
+  using CallbackFn = std::function<void()>;
+
+public:
+  IKafkaStreamDecoder(std::shared_ptr<IKafkaBroker> broker,
+                      const std::string &streamTopic,
+                      const std::string &runInfoTopic,
+                      const std::string &spDetTopic,
+                      const std::string &sampleEnvTopic);
+  virtual ~IKafkaStreamDecoder();
+  IKafkaStreamDecoder(const IKafkaStreamDecoder &) = delete;
+  IKafkaStreamDecoder &operator=(const IKafkaStreamDecoder &) = delete;
+
+public:
+  ///@name Start/stop
+  ///@{
+  void startCapture(bool startNow = true);
+  void stopCapture() noexcept;
+  ///@}
+
+  ///@name Querying
+  ///@{
+  bool isCapturing() const noexcept { return m_capturing; }
+  virtual bool hasData() const noexcept = 0;
+  int runNumber() const noexcept { return m_runNumber; }
+  virtual bool hasReachedEndOfRun() noexcept = 0;
```

Comment:
My understanding of a function declaring itself `noexcept` is that it means:

> I promise not to allow any exceptions to escape my scope, and if I break that promise then **please call std::terminate** for me

This is a runtime feature, not a compile-time check like `const`. Is that really the intent here and the other places it is used?

The only time I'm aware of that this has a significant functional benefit is when marking constructors `noexcept`, which allows libraries like the STL to optimize (re)allocations using things like [move_if_noexcept()](https://en.cppreference.com/w/cpp/utility/move_if_noexcept) to avoid copies.

Other notable properties and side-effects:

- Functions marked `noexcept` can still call functions that aren't `noexcept`, and use exceptions internally
- Only difference is that failing to `catch` one and allowing it to propagate results in `std::terminate`
- Once a virtual function is declared `noexcept`, derived classes cannot override this
- Destructors are always `noexcept` by default, unless explicitly marked `noexcept(false)`

It looks like this in the old KafkaEventStreamDecoder, so not sure this is something that needs to / should be addressed in this PR. Still, something worth considering I think.